### PR TITLE
Block sync initial implementation

### PR DESCRIPTION
**Pull request by netoneko**

### CODE REVIEW

**Comment by netoneko** [comment]
@talkol acceptance test is flaky with `fatal error: concurrent map iteration and map write` inside `TamperingTransport`

File: services/blockstorage/block_sync.go [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+	})
+
+	requestBlocksTrigger := time.AfterFunc(b.config.BlockSyncCollectResponseTimeout(), func() {
+		if state == BLOCK_SYNC_PETITIONER_COLLECTING_AVAILABILITY_RESPONSES {
+			state = BLOCK_SYNC_PETITIONER_ASK_FOR_BLOCKS
+			b.Events <- nil
+		}
+	})
+
+	for {
```
Comment (talkol) [inline services/blockstorage/block_sync.go]: We should extract the for body to a different function that receives the current state and returns the next state and handles the entire body.

This extracted function is the function that we will test in the unit tests directly.

File: services/blockstorage/block_sync.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+		state, blockAvailabilityResponses = b.dispatchEvent(state, event, blockAvailabilityResponses)
+
+		if state == BLOCK_SYNC_STATE_START_SYNC {
+			syncTrigger.Stop()
```
Comment (talkol) [inline services/blockstorage/block_sync.go]: Add a TODO here to handle the timer gotchas. The current `Stop()` and `Reset()` implementation is naive and misses potential edge cases explained in the godocs. @jlevison can help with this since we need a global timer policy for the project

File: services/blockstorage/block_sync.go [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+		if fromMe := message.Sender.SenderPublicKey().Equal(b.config.NodePublicKey()); !fromMe {
+			b.PetitionerHandleBlockSyncResponse(message)
+			return BLOCK_SYNC_STATE_IDLE, blockAvailabilityResponses
+		}
+	}
+
+	return state, blockAvailabilityResponses
+}
+
+func (b *BlockSync) PetitionerBroadcastBlockAvailabilityRequest() error {
```
Comment (talkol) [inline services/blockstorage/block_sync.go]: This method should be private, if it needs to be public for the tests it means that this is a smell and these tests should be converted to unit tests. If it's too complicated to change now, add a TODO and we'll do it once we introduce unit tests

File: services/blockstorage/block_sync.go [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+				LastCommittedBlockHeight: lastCommittedBlockHeight,
+			}).Build(),
+		},
+	}
+
+	_, err := b.gossip.BroadcastBlockAvailabilityRequest(input)
+	return err
+}
+
+func (b *BlockSync) PetitionerHandleBlockSyncResponse(message *gossipmessages.BlockSyncResponseMessage) {
```
Comment (talkol) [inline services/blockstorage/block_sync.go]: This method should be private, if it needs to be public for the tests it means that this is a smell and these tests should be converted to unit tests. If it's too complicated to change now, add a TODO and we'll do it once we introduce unit tests

File: services/blockstorage/block_sync.go [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+
+		_, err = b.storage.CommitBlock(&services.CommitBlockInput{BlockPair: blockPair})
+
+		if err != nil {
+			b.reporting.Error("failed to commit block received via sync", log.Error(err))
+		}
+	}
+}
+
+func (b *BlockSync) PetitionerHandleBlockAvailabilityResponse(message *gossipmessages.BlockAvailabilityResponseMessage) error {
```
Comment (talkol) [inline services/blockstorage/block_sync.go]: same

File: services/blockstorage/block_sync.go [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+	lastCommittedBlockHeight := b.storage.LastCommittedBlockHeight()
+
+	if lastCommittedBlockHeight >= message.SignedBatchRange.LastCommittedBlockHeight() {
+		return errors.New("source is behind petitioner") // stay in the same state
+	}
+
+	return nil
+}
+
+func (b *BlockSync) PetitionerSendBlockSyncRequest(blockType gossipmessages.BlockType, senderPublicKey primitives.Ed25519PublicKey) error {
```
Comment (talkol) [inline services/blockstorage/block_sync.go]: same

File: services/blockstorage/block_sync.go [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+				LastCommittedBlockHeight: lastCommittedBlockHeight,
+			}).Build(),
+		},
+	}
+
+	_, err := b.gossip.SendBlockSyncRequest(request)
+	return err
+}
+
+func (b *BlockSync) SourceHandleBlockAvailabilityRequest(message *gossipmessages.BlockAvailabilityRequestMessage) {
```
Comment (talkol) [inline services/blockstorage/block_sync.go]: same

File: services/blockstorage/service.go [hunk 8]
```diff
@@ -38,16 +43,36 @@ type service struct {
...
+	if err != nil {
+		logger.Error("could not update last block from persistence", log.Error(err))
+	}
+
+	if lastBlock != nil {
+		storage.updateLastCommittedBlock(lastBlock)
+	}
+
+	storage.blockSync = NewBlockSync(ctx, config, storage, gossip, reporting)
+	gossip.RegisterBlockSyncHandler(storage)
```
Comment (talkol) [inline services/blockstorage/service.go]: Change the order of these 2 actions. We should register with gossip first, because the block sync relies on this registration

File: services/blockstorage/service.go [hunk 9]
```diff
@@ -90,7 +115,7 @@ func (s *service) updateLastCommittedBlock(block *protocol.BlockPairContainer) {
 	s.lastCommittedBlock = block
 }
 
-func (s *service) lastCommittedBlockHeight() primitives.BlockHeight {
+func (s *service) LastCommittedBlockHeight() primitives.BlockHeight {
 	if s.lastCommittedBlock == nil {
```
Comment (talkol) [inline services/blockstorage/service.go]: This access needs to be under mutex

File: services/blockstorage/service.go [hunk 10]
```diff
@@ -216,35 +246,41 @@ func (s *service) RegisterConsensusBlocksHandler(handler handlers.ConsensusBlock
 
 	// update the consensus algo about the latest block we have (for its initialization)
 	// TODO: should this be under mutex since it reads s.lastCommittedBlock
+	s.UpdateConsensusAlgo()
+}
+
+func (s *service) UpdateConsensusAlgo() {
 	if s.lastCommittedBlock != nil {
```
Comment (talkol) [inline services/blockstorage/service.go]: Needs to be under mutex

File: services/blockstorage/service.go [hunk 11]
```diff
@@ -297,10 +333,55 @@ func (s *service) validateProtocolVersion(blockPair *protocol.BlockPairContainer
...
-		ContractStateDiffs: commitedBlockPair.ResultsBlock.ContractStateDiffs,
+		ResultsBlockHeader: committedBlockPair.ResultsBlock.Header,
+		ContractStateDiffs: committedBlockPair.ResultsBlock.ContractStateDiffs,
 	})
 	return err
 }
+
+func (s *service) validateWithConsensusAlgos(prevBlockPair *protocol.BlockPairContainer, lastCommittedBlockPair *protocol.BlockPairContainer) error {
+	for _, handler := range s.consensusBlocksHandlers {
+		_, err := handler.HandleBlockConsensus(&handlers.HandleBlockConsensusInput{
```
Comment (talkol) [inline services/blockstorage/service.go]: Once my PR gets merged: https://github.com/orbs-network/orbs-network-go/pull/154
This line needs to be fixed to add the `Mode` argument. Open an issue to remember to do this or merge the other PR first into your PR.

The mode needs to be `VERIFY_AND_UPDATE`

File: services/blockstorage/service.go [hunk 12]
```diff
@@ -297,10 +333,55 @@ func (s *service) validateProtocolVersion(blockPair *protocol.BlockPairContainer
...
+			PrevCommittedBlockPair: prevBlockPair,
+		})
+
+		// one of the consensus algos has validated the block, this means it's a valid block
+		if err == nil {
+			return nil
+		}
+	}
+
+	return errors.New("all consensus algos refused to validate the block")
```
Comment (talkol) [inline services/blockstorage/service.go]: Add count of them (all 3 consensus algos) to avoid issues that some of them did not register yet

File: services/blockstorage/service.go [hunk 13]
```diff
@@ -297,10 +333,55 @@ func (s *service) validateProtocolVersion(blockPair *protocol.BlockPairContainer
...
+}
+
+// Returns a slice of blocks containing first and last
+func (s *service) GetBlocks(first primitives.BlockHeight, last primitives.BlockHeight) (blocks []*protocol.BlockPairContainer, firstAvailableBlockHeight primitives.BlockHeight, lastAvailableBlockHeight primitives.BlockHeight) {
```
Comment (talkol) [inline services/blockstorage/service.go]: Add a TODO to improve this for chunking, now we hold in memory the entire range which is problematic

File: services/blockstorage/test/block_sync_test.go [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+
+	driver.expectCommitStateDiffTimes(2)
+
+	driver.commitBlock(builders.BlockPair().WithHeight(primitives.BlockHeight(1)).WithBlockCreated(time.Now()).Build())
+	driver.commitBlock(builders.BlockPair().WithHeight(primitives.BlockHeight(2)).WithBlockCreated(time.Now()).Build())
+
+	senderKeyPair := keys.Ed25519KeyPairForTests(9)
+
+	input := generateBlockAvailabilityRequestInput(primitives.BlockHeight(0), senderKeyPair.PublicKey())
+	response := &gossiptopics.BlockAvailabilityResponseInput{
```
Comment (talkol) [inline services/blockstorage/test/block_sync_test.go]: Extract this call to a method to make code more readable, this is hard to follow

File: services/blockstorage/test/block_sync_test.go [hunk 15]
```diff
@@ -0,0 +1,441 @@
...
+	_, err := driver.blockStorage.HandleBlockAvailabilityRequest(input)
+	require.NoError(t, err)
+
+	// FIXME remove sleep
+	time.Sleep(1 * time.Millisecond)
+
+	driver.verifyMocks(t)
+}
+
+func TestSyncSourceIgnoresBlockAvailabilityRequestIfPetitionerIsFurtherAhead(t *testing.T) {
```
Comment (talkol) [inline services/blockstorage/test/block_sync_test.go]: We have a very large task of moving many of these tests to unit tests

Let's focus on this task for your next PR

I don't want the next PR to add functionality or features, just convert to unit tests

File: services/blockstorage/test/harness.go [hunk 16]
```diff
@@ -55,18 +81,51 @@ func (d *driver) failNextBlocks() {
 	d.storageAdapter.FailNextBlocks()
 }
 
-func NewDriver() *driver {
```
Comment (talkol) [inline services/blockstorage/test/harness.go]: Why not rename `Driver` to `Harness` to match the rest of the code base? In the acceptance and other services it's always named harness

File: services/blockstorage/test/init_test.go [hunk 17]
```diff
@@ -0,0 +1,49 @@
+package test
```
Comment (talkol) [inline services/blockstorage/test/init_test.go]: Add a test here that tests the registration for the gossip handler, it's part of the spec and not tested

File: services/gossip/topic_block_sync.go [hunk 18]
```diff
@@ -1,20 +1,200 @@
...
+		})
+	}
+}
+
 func (s *service) SendBlockAvailabilityResponse(input *gossiptopics.BlockAvailabilityResponseInput) (*gossiptopics.EmptyOutput, error) {
-	panic("Not implemented")
+	header := (&gossipmessages.HeaderBuilder{
+		Topic:         gossipmessages.HEADER_TOPIC_BLOCK_SYNC,
+		BlockSync:     gossipmessages.BLOCK_SYNC_AVAILABILITY_RESPONSE,
+		RecipientMode: gossipmessages.RECIPIENT_LIST_MODE_BROADCAST,
```
Comment (talkol) [inline services/gossip/topic_block_sync.go]: This should not be broadcast!

File: services/gossip/topic_block_sync.go [hunk 19]
```diff
@@ -1,20 +1,200 @@
...
+	}).Build()
+
+	if input.Message.SignedBatchRange == nil {
+		return nil, errors.Errorf("cannot encode BlockAvailabilityResponseMessage", log.Stringable("message", input.Message))
+	}
+	payloads := [][]byte{header.Raw(), input.Message.SignedBatchRange.Raw(), input.Message.Sender.Raw()}
+
+	return nil, s.transport.Send(&adapter.TransportData{
+		SenderPublicKey: s.config.NodePublicKey(),
+		RecipientMode:   gossipmessages.RECIPIENT_LIST_MODE_BROADCAST,
```
Comment (talkol) [inline services/gossip/topic_block_sync.go]: This should not be broadcast

File: services/gossip/topic_block_sync.go [hunk 20]
```diff
@@ -1,20 +1,200 @@
...
+	if len(payloads) < 3 {
+		return
+	}
+	chunkRange := gossipmessages.BlockSyncRangeReader(payloads[0])
+	senderSignature := gossipmessages.SenderSignatureReader(payloads[1])
+	blockPairPayloads := payloads[2:]
+
+	var blocks []*protocol.BlockPairContainer
+
+	for i := 0; i < len(blockPairPayloads); i += 8 {
```
Comment (talkol) [inline services/gossip/topic_block_sync.go]: ahhhh, you can't use 8 as a hard coded value

this number depends on the number of transactions, receipts and state diffs in the block!

you have to have a dynamic calculation

File: test/acceptance/block_sync_test.go [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+		for i := 1; i <= 10; i++ {
+			blockPair := builders.BlockPair().WithHeight(primitives.BlockHeight(i)).Build()
+			network.BlockPersistence(0).WriteBlock(blockPair)
+		}
+	}).Start(func(network harness.AcceptanceTestNetwork) {
+		if err := network.BlockPersistence(0).GetBlockTracker().WaitForBlock(10); err != nil {
+			t.Errorf("waiting for block on node 0 failed: %s", err)
+		}
+
+		if err := network.BlockPersistence(1).GetBlockTracker().WaitForBlock(5); err != nil {
```
Comment (talkol) [inline test/acceptance/block_sync_test.go]: Redundant, we wait in the next line until we reach 10

File: test/acceptance/block_sync_test.go [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+			t.Errorf("waiting for block on node 1 failed: %s", err)
+		}
+
+		// Wait until full sync
+		if err := network.BlockPersistence(1).GetBlockTracker().WaitForBlock(10); err != nil {
+			t.Errorf("waiting for block on node 1 failed: %s", err)
+		}
+
+		// Wait again to get new blocks created after the sync
```
Comment (talkol) [inline test/acceptance/block_sync_test.go]: We might want to do this for the leader as well

File: services/blockstorage/block_sync.go [hunk 23]
```diff
@@ -0,0 +1,323 @@
...
+	Events  chan interface{}
+}
+
+func NewBlockSync(ctx context.Context, config Config, storage BlockSyncStorage, gossip gossiptopics.BlockSync, reporting log.BasicLogger) *BlockSync {
+	blockSync := &BlockSync{
+		reporting: reporting.For(log.String("flow", "block-sync")),
+		config:    config,
+		storage:   storage,
+		gossip:    gossip,
+		Events:    make(chan interface{}),
```
[the remark this answers is not in the document]
Comment (netoneko) [reply services/blockstorage/block_sync.go]: I thought that I won't be able to access it from block storage service, but turns out I can. Fixed.

File: services/blockstorage/block_sync.go [hunk 24]
```diff
@@ -0,0 +1,323 @@
...
+	syncTrigger := time.AfterFunc(b.config.BlockSyncInterval(), func() {
+		if state == BLOCK_SYNC_STATE_IDLE {
+			state = BLOCK_SYNC_STATE_START_SYNC
+			b.Events <- nil
+		}
+	})
+
+	requestBlocksTrigger := time.AfterFunc(b.config.BlockSyncCollectResponseTimeout(), func() {
+		if state == BLOCK_SYNC_PETITIONER_COLLECTING_AVAILABILITY_RESPONSES {
+			state = BLOCK_SYNC_PETITIONER_ASK_FOR_BLOCKS
```
[the remark this answers is not in the document]
Comment (netoneko) [reply services/blockstorage/block_sync.go]: Fixed.

File: services/blockstorage/block_sync.go (hunk 3 above)
Comment (netoneko) [reply services/blockstorage/block_sync.go]: Fixed

File: services/blockstorage/block_sync.go (hunk 5 above)
Comment (netoneko) [reply services/blockstorage/block_sync.go]: Fixed

File: services/blockstorage/block_sync.go (hunk 7 above)
Comment (netoneko) [reply services/blockstorage/block_sync.go]: Fixed

File: services/blockstorage/block_sync.go (hunk 6 above)
Comment (netoneko) [reply services/blockstorage/block_sync.go]: Fixed

File: services/blockstorage/service.go (hunk 12 above)
Comment (netoneko) [reply services/blockstorage/service.go]: Fixed

File: services/blockstorage/service.go (hunk 13 above)
Comment (netoneko) [reply services/blockstorage/service.go]: Fixed

File: services/blockstorage/block_sync.go (hunk 2 above)
Comment (netoneko) [reply services/blockstorage/block_sync.go]: Fixed

File: services/blockstorage/block_sync.go (hunk 4 above)
Comment (netoneko) [reply services/blockstorage/block_sync.go]: Fixed

File: services/blockstorage/test/harness.go (hunk 16 above)
Comment (netoneko) [reply services/blockstorage/test/harness.go]: Done.

File: test/acceptance/block_sync_test.go (hunk 21 above)
Comment (talkol) [reply test/acceptance/block_sync_test.go]: ok

File: test/acceptance/block_sync_test.go (hunk 22 above)
Comment (talkol) [reply test/acceptance/block_sync_test.go]: implementation detail you're not supposed to know :)

File: services/blockstorage/test/block_sync_test.go (hunk 14 above)
Comment (netoneko) [reply services/blockstorage/test/block_sync_test.go]: Fixed.

File: services/blockstorage/block_sync.go (hunk 1 above)
Comment (netoneko) [reply services/blockstorage/block_sync.go]: I don't think it's possible now because it requires another refactoring. I will do it in another pr.

File: services/blockstorage/block_sync.go [hunk 25]
```diff
@@ -0,0 +1,323 @@
...
+	return blockSync
+}
+
+func (b *BlockSync) mainLoop(ctx context.Context) {
+	state := BLOCK_SYNC_STATE_START_SYNC
+	var event interface{}
+	var blockAvailabilityResponses []*gossipmessages.BlockAvailabilityResponseMessage
+
+	syncTrigger := time.AfterFunc(b.config.BlockSyncInterval(), func() {
+		if state == BLOCK_SYNC_STATE_IDLE {
```
[the remark this answers is not in the document]
Comment (netoneko) [reply services/blockstorage/block_sync.go]: Fixed.

File: services/gossip/topic_block_sync.go (hunk 19 above)
Comment (netoneko) [reply services/gossip/topic_block_sync.go]: Fixed.

File: services/gossip/topic_block_sync.go (hunk 18 above)
Comment (netoneko) [reply services/gossip/topic_block_sync.go]: Fixed.

File: services/blockstorage/test/init_test.go (hunk 17 above)
Comment (netoneko) [reply services/blockstorage/test/init_test.go]: We have this mock in every harness setup.

File: services/blockstorage/service.go (hunk 8 above)
Comment (netoneko) [reply services/blockstorage/service.go]: Fixed.

File: services/blockstorage/service.go (hunk 9 above)
Comment (netoneko) [reply services/blockstorage/service.go]: Fixed.

File: services/blockstorage/service.go (hunk 10 above)
Comment (netoneko) [reply services/blockstorage/service.go]: Fixed.

File: services/gossip/topic_block_sync.go (hunk 20 above)
Comment (netoneko) [reply services/gossip/topic_block_sync.go]: Fixed
